src/code/padmgr.c

```diff
@@ -23,69 +76,94 @@ OSMesgQueue* PadMgr_LockSerialMesgQueue(PadMgr* padMgr) {
         osSyncPrintf("%2d %d serialMsgQロック解除します   %08x %08x %08x\n", osGetThreadId(NULL),
                      MQ_GET_COUNT(&padMgr->serialLockQueue), padMgr, &padMgr->serialLockQueue, serialEventQueue);
     }
 
     osSendMesg(&padMgr->serialLockQueue, (OSMesg)serialEventQueue, OS_MESG_BLOCK);
 
-    if (D_8012D280 > 2) {
+    if (gPadMgrLogSeverity >= LOG_SEVERITY_VERBOSE) {
         // "serialMsgQ Unlocked"
         osSyncPrintf("%2d %d serialMsgQロック解除しました %08x %08x %08x\n", osGetThreadId(NULL),
                      MQ_GET_COUNT(&padMgr->serialLockQueue), padMgr, &padMgr->serialLockQueue, serialEventQueue);
     }
 }
 
+/**
+ * Locks controller input data while padmgr is reading new inputs or another thread is using the current inputs.
+ * This prevents new inputs overwriting the current inputs while they are in use.
+ *
+ * @see PadMgr_UnlockPadData
+ */
 void PadMgr_LockPadData(PadMgr* padMgr) {
     osRecvMesg(&padMgr->lockQueue, NULL, OS_MESG_BLOCK);
 }
 
+/**
+ * Unlocks controller input data, allowing padmgr to read new inputs or another thread to access the currently
+ * read inputs.
+ *
+ * @see PadMgr_LockPadData
+ */
 void PadMgr_UnlockPadData(PadMgr* padMgr) {
     osSendMesg(&padMgr->lockQueue, NULL, OS_MESG_BLOCK);
 }
 
-void PadMgr_RumbleControl(PadMgr* padMgr) {
-    static u32 errcnt = 0;
-    static u32 frame;
-    s32 temp = 1;
+/**
+ * Activates the rumble pak for all controllers it is enabled on, stops it for all controllers it is disabled on and
+ * attempts to initialize it for a controller if it is not already initialized.
+ */
+void PadMgr_UpdateRumble(PadMgr* padMgr) {
+    static u32 sRumbleErrorCount = 0; // original name: "errcnt"
+    static u32 sRumbleUpdateCounter;
+    s32 one = 1; // Required for matching
     s32 triedRumbleComm;
-    OSMesgQueue* serialEventQueue = PadMgr_LockSerialMesgQueue(padMgr);
-    s32 var4;
+    OSMesgQueue* serialEventQueue = PadMgr_AcquireSerialEventQueue(padMgr);
+    s32 ret;
     s32 i;
 
-    triedRumbleComm = 0;
+    triedRumbleComm = false;
 
-    for (i = 0; i < 4; i++) {
+    for (i = 0; i < MAXCONTROLLERS; i++) {
         if (padMgr->ctrlrIsConnected[i]) {
-            if (padMgr->padStatus[i].status & 1) {
-                if (padMgr->pakType[i] == temp) {
-                    if (padMgr->rumbleEnable[i] != 0) {
-                        if (padMgr->rumbleCounter[i] < 3) {
+            // Check status for whether a controller pak is connected
+            if (padMgr->padStatus[i].status & CONT_CARD_ON) {
+                if (padMgr->pakType[i] == one) { // CONT_PAK_RUMBLE
+                    if (padMgr->rumbleEnable[i]) {
+                        if (padMgr->rumbleTimer[i] < 3) {
                             // clang-format off
                             if (1) {} osSyncPrintf(VT_FGCOL(YELLOW));
                             // clang-format on
 
-                            // "Vibration pack jumble jumble"?
+                            // "Rumble pack brrr"
```

Comment:
Looks like this also has info on which controller is rumbling, `コン` being short for "controller"
```suggestion
                            // "Controller `i+1` Rumble pack brrr"
```
or similar.